dds/DCPS/transport/framework/NetworkAddress.cpp

```diff
@@ -473,6 +473,87 @@ bool open_appropriate_socket_type(ACE_SOCK_Dgram& socket, const ACE_INET_Addr& l
   return socket.open(local_address) == 0;
 #endif
 }
+
+ACE_INET_Addr choose_single_coherent_address(const ACE_INET_Addr& address)
+{
+#if ACE_MAJOR_VERSION < 6 || (ACE_MAJOR_VERSION == 6 && (ACE_MINOR_VERSION < 3 || (ACE_MINOR_VERSION == 3 && ACE_MICRO_VERSION < 1)))
+  // ACE_INET_Addr does not support next()
+  return address;
+#else
+  ACE_INET_Addr copy(address);
+
+#ifdef ACE_HAS_IPV6
+  OPENDDS_SET(ACE_INET_Addr) set6;
+#endif
+  OPENDDS_SET(ACE_INET_Addr) set4;
+
+  do {
+    ACE_INET_Addr temp;
+    temp.set_addr(copy.get_addr(), copy.get_addr_size());
+
+#ifdef ACE_HAS_IPV6
+    if (temp.get_type() == AF_INET6 && !temp.is_multicast()) {
+      if (!temp.is_ipv4_mapped_ipv6() && !temp.is_ipv4_compat_ipv6()) {
+        set6.insert(temp);
+      } else {
+        set4.insert(temp);
+      }
+    }
+#endif
+    if (temp.get_type() == AF_INET && !temp.is_multicast()) {
+      set4.insert(temp);
+    }
+  } while (copy.next());
+
+#ifdef ACE_HAS_IPV6
+  ACE_INET_Addr loop6;
+  OPENDDS_SET(ACE_INET_Addr) set6_local;
+  for (OPENDDS_SET(ACE_INET_Addr)::const_iterator it = set6.begin(); it != set6.end(); /* inc in loop */) {
+    if (it->is_loopback()) {
+      loop6 = *it;
+      set6.erase(it++);
+    } else if (it->is_linklocal()) {
+      set6_local.insert(*it);
+      set6.erase(it++);
+    } else {
+      ++it;
+    }
+  }
+#endif
+
+  ACE_INET_Addr loop4;
+  for (OPENDDS_SET(ACE_INET_Addr)::const_iterator it = set4.begin(); it != set4.end(); /* inc in loop */) {
+    if (it->is_loopback()) {
+      loop4 = *it;
+      set4.erase(it++);
+    } else {
+      ++it;
+    }
+  }
+
+#ifdef ACE_HAS_IPV6
+  if (loop6 != ACE_INET_Addr()) {
+    return loop6;
```

Comment:
Why is loop preferred over non-loop?

Reply:
If one of the resolved addresses for the hostname in question is loopback, we're asking about ourselves, so just use loopback.

I am not super tied to any of the logic in this method other than sorting, I was just trying to pick something somewhat sensible to start with.

In practice, I am not sure INET_Addr is even this sophisticated. I never saw a mix of IPv4 and IPv6 in my testing, but I saw a IPv4 mapped IPv6 address that didn't seem to get flagged appropriately by that API. So in reality, maybe everything other than sorting can go away?

Reply:
The loopback aspect that concerns me is that it's used for `TcpInst::local_address_` which seems like that could impact what the socket binds to.

Reply:
I originally thought about adding an "allow_loopback" flag to the arguments for use with tcp & udp transports' local addresses, but in testing it never seemed to be a problem. I think this is partly because it's rare for hosts to resolve any hostname (including their own) to the loopback address, so I left it out. I can either add the flag and apply it for local addresses, or I can remove loopback addresses entirely.